### ISSUE TITLE
[Bug]: GDN Phase6 final_state 精度在尾块仅含 1–15 个有效 token 时显著下降

### ISSUE THREAD

**Issue by wanghuanjun2113** [flashserve/flash-linear-attention-npu#407]
## 环境信息

- 硬件：Ascend 910B3
- CANN：8.5.1
- Python：3.11.14
- PyTorch：2.9.0
- torch_npu：2.9.0.post7
- fla_npu：26.7.0.dev0
- 复现版本：fork yjmyl/flash-linear-attention-npu 的 chw 分支
- commit：[e0fb894](https://github.com/yjmyl/flash-linear-attention-npu/commit/e0fb8942336efcf3b2e13d7463e63b5b5d38f341)
- dtype/layout：BF16 Q/K/V，FP32 g 与 initial_state，head-first Q/K/V，varlen metadata
- shape：Hk=4, Hv=12, K=V=128, chunk_size=64

## 问题概述

gdn_core_fwd_phase6 在最后一个 64-token chunk 只有 **1–15 个有效 token** 时，返回的 final_state 与参考实现产生约 3e-2 到 5e-2 的最大绝对误差。

同一次调用返回的注意力输出 o 仍与参考实现一致，max_abs 稳定在约 3.05e-5。问题只集中在 final_state，因此短请求可能看不出异常，但该 state 被后续 decode、prefix cache 或后续 chunk 使用时会影响模型行为。

在 e0fb894 上，同输入重复执行三次，错误结果 bit-exact，因此这里观察到的是**确定性的尾块精度问题**，不是随机抖动。

参考实现使用 vLLM-Ascend Triton chunk_gated_delta_rule，并启用相同的 Q/K L2 normalization。

## 最小复现

~~~python
import torch
from fla_npu.ops import ascendc
from vllm_ascend.ops.triton.fla.chunk import (
    chunk_gated_delta_rule as reference_gdn,
)
from vllm_ascend.ops.triton.fla.l2norm import l2norm_fwd

torch.manual_seed(1001)
torch.npu.set_device(0)

Hk, Hv, D = 4, 12, 128
T = 1281  # T % 64 == 1; 1..15 均可复现
dtype = torch.bfloat16
cu = [0, T]

q = torch.randn(1, T, Hk, D, device="npu", dtype=dtype) * 0.05
k = torch.randn(1, T, Hk, D, device="npu", dtype=dtype) * 0.05
v = torch.randn(1, T, Hv, D, device="npu", dtype=dtype) * 0.05
g = -torch.rand(1, T, Hv, device="npu", dtype=torch.float32) * 0.1
beta = torch.sigmoid(torch.randn(1, T, Hv, device="npu", dtype=dtype))
initial_state = (
    torch.randn(1, Hv, D, D, device="npu", dtype=torch.float32) * 0.01
)

cu_tensor = torch.tensor(cu, device="npu", dtype=torch.int32)
o_ref, state_ref = reference_gdn(
    q=q, k=k, v=v, g=g, beta=beta,
    initial_state=initial_state.clone(),
    output_final_state=True,
    cu_seqlens=cu_tensor,
    use_qk_l2norm_in_kernel=True,
)

q_head = l2norm_fwd(q).transpose(1, 2).contiguous()
k_head = l2norm_fwd(k).transpose(1, 2).contiguous()
v_head = v.transpose(1, 2).contiguous()
chunk_indices = tuple(
    x for chunk_id in range((T + 63) // 64) for x in (0, chunk_id)
)

o_fused, state_fused, _, _ = ascendc.gdn_core_fwd_phase6(
    q_head, k_head, v_head, g, beta.float(),
    initial_state=initial_state.clone(),
    output_final_state=True,
    chunk_size=64,
    cu_seqlens=tuple(cu),
    chunk_indices=chunk_indices,
    scale=D**-0.5,
)
o_fused = o_fused.transpose(1, 2)
torch.npu.synchronize()

print("o max_abs:",
      (o_fused.float() - o_ref.float()).abs().max().item())
print("state max_abs:",
      (state_fused - state_ref).abs().max().item())
~~~

## 实测边界矩阵

固定主体长度为 1280 token，只改变最后一个 chunk 的有效 token 数：

| T | T % 64 | final_state max_abs | o max_abs |
|---:|---:|---:|---:|
| 1280 | 0 | 2.30e-4 | 3.05e-5 |
| 1281 | 1 | 3.33e-2 | 3.05e-5 |
| 1282 | 2 | 4.08e-2 | 3.05e-5 |
| 1283 | 3 | 3.49e-2 | 3.05e-5 |
| 1284 | 4 | 4.35e-2 | 3.05e-5 |
| 1285 | 5 | 3.79e-2 | 3.05e-5 |
| 1286 | 6 | 3.70e-2 | 3.05e-5 |
| 1287 | 7 | 3.82e-2 | 3.05e-5 |
| 1288 | 8 | 4.19e-2 | 3.05e-5 |
| 1289 | 9 | 4.95e-2 | 3.05e-5 |
| 1295 | 15 | 4.88e-2 | 3.05e-5 |
| 1296 | 16 | 2.16e-4 | 3.05e-5 |
| 1312 | 32 | 2.52e-4 | 4.58e-5 |
| 1343 | 63 | 2.67e-4 | 3.05e-5 |

其他代表点：

| 场景 | final_state max_abs |
|---|---:|
| 单序列 T=1000，余数40 | 2.68e-4 |
| 单序列 T=1500，余数28 | 1.85e-4 |
| 单序列 T=1347，余数3 | 3.66e-2 |
| 单序列 T=3847，余数7 | 4.50e-2 |
| varlen cu=[0,1000,2500,3847] | 3.73e-2 |

最后一个 varlen case 的逐序列 state 误差为：

~~~text
[3.07e-4, 3.07e-4, 3.73e-2]
~~~

只有尾长为 3 的第三条序列失真，说明问题由每条序列自身的尾块长度决定，而不是 packed-varlen 的全局 token 位置决定。

## 无异常栈，表现为数值错误

算子调用正常完成，没有抛出异常，所有输出均 finite：

~~~text
o max_abs:     approximately 3.05e-5
state max_abs: approximately 3e-2 to 5e-2
~~~

## no-op padding 对照实验

调用融合算子前，仅将尾长 1–15 的序列补到 16 token：

- q/k/v = 0
- beta = 0，不产生 delta-rule update
- g = 0，不增加额外衰减
- 调用后裁掉补位位置的 o

结果：

| 场景 | o max_abs | final_state max_abs |
|---|---:|---:|
| T=1283，补到1296 | 3.05e-5 | 2.97e-4 |
| cu=[0,1000,2500,3847]，最后一条补13 token | 3.05e-5 | 2.79e-4 |
| 三条序列尾长均为1–15，分别补到16 | 3.05e-5 | 2.95e-4 |

这说明数学主路径可以保持精度，问题与小于 16 token 的专用 tail fallback 高度相关。

## 相关代码边界与已排除项

gdn_fwd_h_kernel.hpp 在 blockTokens 小于 16 时会绕过 cube GEMM，转到：

- ComputeTailVWorkspace
- ComputeTailHWorkspace

建议重点检查：

1. 小尾块 V workspace 的生产与 H workspace 的消费顺序；
2. 两个 AIV subblock 的共享 workspace 可见性；
3. 小于 16 fallback 的 state update 与正常 cube path 是否完全等价；
4. FP32 initial_state/final_state 路径中的事件复用。

曾尝试在 tail path 发布 vec1Done 前增加 AIV subblock barrier，但 state 误差没有改善，该尝试已撤回。因此目前不能把问题简单归因于 vec1Done 发布过早。

## 期望行为

对于任意合法尾块长度 1–63：

- final_state 精度应与尾长 0 或大于等于 16 时保持同一量级；
- 对本组 BF16 输入，期望 final_state max_abs 约为 1e-4，而不是 1e-2；
- o 与 final_state 应同时通过与非融合参考路径的精度门禁。

建议增加回归矩阵：

~~~text
tail_tokens = [1, 2, 3, 7, 8, 15, 16, 17, 31, 32, 63, 64]
initial_state dtype = [BF16, FP32]
layout = [dense, varlen]
sequence_count = [1, 3]
~~~
